### PR TITLE
[TypeInfo] Redesign Type methods and nullability

### PR DESCRIPTION
**Pull request by mtarld**
| Q             | A
| ------------- | ---
| Branch?       | 7.2
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        |
| License       | MIT

By using the component and trying to integrate it within several libs, I ended up finding some limitations due to the component design.

Basically, the issues were:
- Abstract classes (such as `Type`) knowing its concrete implementations (such as in `getBaseType` method).
- For API simplicity `CollectionType`, and `GenericType` are forwarding methods to their wrapped type using `__call`, but I think this must be avoided as much as possible.
- It is difficult to know if a type is "just" a nullable type (because we need to check the union content).
- Invalid types could be instantiated

To tackle these issues, I propose the following:
- Introduce `CompositeTypeInterface` and `WrappingTypeInterface` that describe whether a type is composed by several types or wrapping one.
- Add a `NullableType` which extends `UnionType` and is a `WrappingTypeInterface`, so that it is easy to recognize a nullable type and to get the non-nullable related type.
- Remove the `Type::asNonNullable` method as it's covered by `NullableType::getWrappedType`
- Get rid of the `__call` method as it is easy to know it a type is wrapping another one
- Rename `is` to `satisfy` and add the new `CompositeTypeInterface::composedTypesSatisfy` and `WrappingTypeInterface::wrappedTypeSatisfy` methods
- Rename `isA` to `isIdentifiedBy`

Plus, **thanks to @skalpa's great work**, the following checks are added to prevent invalid type construction:
- `BackedEnumType`
    - Backed type can only be `int` or `string`
- `CollectionType`
    - Main type can only be `array` or `iterable` when it's a built-in type
- `GenericType`
    - Main type can only be `array` or `iterable` when it's a built-in type
- `IntersectionType`
    - Cannot contain other types than class types
- `UnionType`
    - Cannot contain `UnionType`
    - Cannot contain standalone types (`mixed`, `void`, and `never`)
    - Cannot contain either `true` and `false`
    - Cannot contain either `bool` and `true` or `false`
    - Cannot contain either `object` and a class type

### CODE REVIEW

**Comment by stof** [comment]
Using public properties instead of getters makes the maintenance harder in case we need to make the API of the component evolve (as we cannot rely on property hooks yet to implement BC layers)

File: src/Symfony/Component/PropertyInfo/Extractor/PhpDocExtractor.php [hunk 1]
```diff
@@ -11,6 +11,7 @@
 
 namespace Symfony\Component\PropertyInfo\Extractor;
 
+use Symfony\Component\TypeInfo\Specification\IsNullable;
```
Comment (stof) [inline src/Symfony/Component/PropertyInfo/Extractor/PhpDocExtractor.php]: this seems unused

File: src/Symfony/Component/TypeInfo/Specification/IsA.php [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+{
+    /**
+     * @param TypeIdentifier|class-string $expected
+     */
+    public function __construct(
+        private TypeIdentifier|string $expected,
+    ) {
+    }
+
+    public function isSatifiedBy(Type $type): bool
```
Comment (stof) [inline src/Symfony/Component/TypeInfo/Specification/IsA.php]: There is a big drawback of this proposal: the implementation of this method assumes that it knows all the possible implementations of Type (as it has to handle them in different way). This assumption is wrong (nothing forbids someone to create another class extending Type)

Comment (mtarld) [reply src/Symfony/Component/TypeInfo/Specification/IsA.php]: Indeed, you were totally right, I moved back to the "abstract class" methods but in a way that ignores the concrete implementations.

File: src/Symfony/Component/PropertyInfo/Util/PhpDocTypeHelper.php [hunk 3]
```diff
@@ -128,7 +128,9 @@ public function getType(DocType $varType): ?Type
                 $nullable = true;
             }
 
-            return $this->createType($varType, $nullable);
+            $type = $this->createType($varType, $nullable);
```
Comment (jdreesen) [inline src/Symfony/Component/PropertyInfo/Util/PhpDocTypeHelper.php]:
```suggestion
            $type = $this->createType($varType);
```

Comment (mtarld) [reply src/Symfony/Component/PropertyInfo/Util/PhpDocTypeHelper.php]: Nice catch! Thanks :slightly_smiling_face:

**Review by Korbeil — APPROVED** [review]
Loving all the changes ! :rocket:

File: src/Symfony/Bridge/Doctrine/composer.json [hunk 4]
```diff
@@ -39,7 +39,7 @@
         "symfony/security-core": "^6.4|^7.0",
         "symfony/stopwatch": "^6.4|^7.0",
         "symfony/translation": "^6.4|^7.0",
-        "symfony/type-info": "^7.1",
+        "symfony/type-info": "^7.2",
```
Comment (xabbuh) [inline src/Symfony/Bridge/Doctrine/composer.json:42]: Is this version bump needed?

File: src/Symfony/Bundle/FrameworkBundle/composer.json [hunk 5]
```diff
@@ -64,7 +64,7 @@
         "symfony/string": "^6.4|^7.0",
         "symfony/translation": "^6.4|^7.0",
         "symfony/twig-bundle": "^6.4|^7.0",
-        "symfony/type-info": "^7.1",
+        "symfony/type-info": "^7.2",
```
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/composer.json:67]: Is this version bump needed?

File: src/Symfony/Component/TypeInfo/Type/CompositeTypeInterface.php [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\TypeInfo\Type;
+
+use Symfony\Component\TypeInfo\Type;
+
+/**
+ * Represents a type composed by several other types.
```
Comment (xabbuh) [inline src/Symfony/Component/TypeInfo/Type/CompositeTypeInterface.php]:
```suggestion
 * Represents a type composed of several other types.
```

File: src/Symfony/Component/TypeInfo/Type/TemplateType.php [hunk 7]
```diff
@@ -27,33 +25,18 @@ final class TemplateType extends Type
...
         return $this->name;
     }
 
-    public function getBound(): Type
-    {
-        return $this->bound;
-    }
-
-    public function asNonNullable(): self
+    public function getBoundType(): Type
```
Comment (xabbuh) [inline src/Symfony/Component/TypeInfo/Type/TemplateType.php]: the renaming from `getBound()` to `getBoundType()` should also be documented in the changelog file

File: src/Symfony/Component/TypeInfo/Type.php [hunk 8]
```diff
@@ -25,35 +24,70 @@ abstract class Type implements \Stringable
...
+            }
+        }
+
+        return false;
     }
 
+    /**
+     * Tells if the type or one of its component (@see traverse()) is nullable.
+     *
+     * @param TypeIdentifier|class-string $expected
```
Comment (stof) [inline src/Symfony/Component/TypeInfo/Type.php]: wrong phpdoc

File: src/Symfony/Component/TypeInfo/Type.php [hunk 9]
```diff
@@ -25,35 +24,70 @@ abstract class Type implements \Stringable
...
+     * Tells if the type or one of its component (@see traverse()) is a scalar.
      *
-     * @param list<mixed> $arguments
+     * @param TypeIdentifier|class-string $expected
      */
-    public function __call(string $method, array $arguments): mixed
+    public function isScalar(): bool
     {
-        throw new LogicException(\sprintf('Cannot call "%s" on "%s" type.', $method, $this));
+        return $this->isA(TypeIdentifier::INT)
```
Comment (stof) [inline src/Symfony/Component/TypeInfo/Type.php]: instead of separate `isA` calls that will require separate traversals, we should perform a single traversing using a combined check for all identifiers (maybe with an API inspired by `is` taking a callback with a type)

File: src/Symfony/Component/TypeInfo/Type.php [hunk 10]
```diff
@@ -25,35 +24,70 @@ abstract class Type implements \Stringable
...
+     * Tells if the type or one of its component (@see traverse()) is what is expected.
+     *
+     * @param TypeIdentifier|class-string $expected
      */
-    public function is(callable $callable): bool
+    public function isA(TypeIdentifier|string $expected): bool
     {
-        return $callable($this);
+        foreach ($this->traverse() as $type) {
+            if ($type !== $this && $type->isA($expected)) {
```
Comment (stof) [inline src/Symfony/Component/TypeInfo/Type.php]: Wouldn't that perform multiple traversals of parts of the object graph when one of the nested type is itself a composite or a wrapping type, and this type does not override `isA` ? `traverse` does not yield only leaf nodes

File: src/Symfony/Component/TypeInfo/Type/NullableType.php [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+final class NullableType extends UnionType implements WrappingTypeInterface
+{
+    /**
+     * @param T $type
+     */
+    public function __construct(
+        private readonly Type $type,
+    ) {
+        if ($type instanceof UnionType) {
+            parent::__construct(Type::null(), ...$type->getTypes());
```
Comment (stof) [inline src/Symfony/Component/TypeInfo/Type/NullableType.php:42]: should we really perform this unwrapping here ?

File: src/Symfony/Component/TypeInfo/Type/NullableType.php [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+
+            return;
+        }
+
+        parent::__construct(Type::null(), $type);
+    }
+
+    public function getWrappedType(): Type
```
Comment (stof) [inline src/Symfony/Component/TypeInfo/Type/NullableType.php:50]: this type actually composes both this type and the NULL type. It might cause confusion for code not handling properly types implementing both WrappingTypeInterface and IntersectionType.

And this `getWrappedType` is not even a replacement for `asNonNullable` as nothing prevents the NULL type to be part of the wrapped type.

File: src/Symfony/Component/TypeInfo/Type.php [hunk 13]
```diff
@@ -25,35 +24,70 @@ abstract class Type implements \Stringable
...
+            yield from $this->getWrappedType()->traverse();
+        }
 
-    abstract public function asNonNullable(): self;
+        yield $this;
+    }
 
     /**
-     * @param callable(Type): bool $callable
+     * Tells if the type or one of its component (@see traverse()) is what is expected.
```
Comment (stof) [inline src/Symfony/Component/TypeInfo/Type.php]: This `isA` API is confusing IMO, because it does not allow to determine whether a type is either a subtype nor a supertype of another type.
Common questions related to types would be "does this type accept this kind of value?", which cannot be answered by this API (the best proof is the fact that both UnionType and IntersectionType uses this same implementation in the base class).
The `is` method that you removed in this PR *was* able to answer such question.

File: src/Symfony/Component/TypeInfo/Type/BuiltinType.php [hunk 14]
```diff
@@ -46,43 +40,13 @@ public function getTypeIdentifier(): TypeIdentifier
...
-     * @return self|UnionType<BuiltinType<TypeIdentifier::OBJECT>|BuiltinType<TypeIdentifier::RESOURCE>|BuiltinType<TypeIdentifier::ARRAY>|BuiltinType<TypeIdentifier::STRING>|BuiltinType<TypeIdentifier::FLOAT>|BuiltinType<TypeIdentifier::INT>|BuiltinType<TypeIdentifier::BOOL>>
-     */
-    public function asNonNullable(): self|UnionType
-    {
-        if (TypeIdentifier::NULL === $this->typeIdentifier) {
-            throw new LogicException('"null" cannot be turned as non nullable.');
-        }
-
-        // "mixed" is an alias of "object|resource|array|string|float|int|bool|null"
-        // therefore, its non-nullable version is "object|resource|array|string|float|int|bool"
```
Comment (stof) [inline src/Symfony/Component/TypeInfo/Type/BuiltinType.php:72]: the removal of this logic is another proof that your assumption that making a non-nullable is simple without a dedicated API is wrong.

File: src/Symfony/Component/TypeInfo/Type/TemplateType.php [hunk 15]
```diff
@@ -27,33 +25,18 @@ final class TemplateType extends Type
...
-    {
-        return false;
-    }
-
     public function getName(): string
     {
         return $this->name;
     }
 
-    public function getBound(): Type
```
Comment (stof) [inline src/Symfony/Component/TypeInfo/Type/TemplateType.php:48]: is the renaming of this method really worth a BC break ? Experimental components allow BC breaks outside major versions, but we still try to maintain BC as much as possible to make upgrades easier.

**Review by stof — COMMENTED** [review]
In the current state of this PR, I'm voting -1 as it reduces the power of the component.

File: src/Symfony/Component/TypeInfo/Type.php (hunk 13 above)
Comment (skalpa) [reply src/Symfony/Component/TypeInfo/Type.php]: @stof. Agreed, that is exactly one of my use case:  I need to determine whether a particular value is valid where a specific type-hint is used and this PR would render this method useless to me.
However, the current implementation has issues nevertheless: on one side `Type::object('SomeClass')->isA('SomeInterface')` but on the other `Type::union(Type::int(), Type::string())->isA(TypeIdentifier::INT)`: it's not clear whether the type is supposed to accept what is passed to isA() or if it is the other way around. 
I was actually going to work on type comparisons, I'd suggest leaving `isA()` untouched for now and reviewing/improving the current implementation separately.

**Comment by skalpa** [comment]
> It is difficult to know if a type is "just" a nullable type (because we need to check the union content).

I am not sure I understand what problem your are trying to solve. A type is "just" a nullable type if `$type instanceof BuiltinType && TypeIdentifier::NULL === $type->getTypeIdentifier()`.

Nevertheless, the changes feel a lot complex considering nullability is not a generic problem. There are only 3 kinds of nullable types: `null`, `mixed`, and unions that contain `null`.

Therefore, I would only keep the current `isNullable()`, and `asNonNullable()` methods, but refactor them a bit:

* The implementation of `Type::isNullable()` should just return `false`
* `Type::asNonNullable()` should be implemented and just return `$this`
* All the unnecessary `asNonNullable()` in children classes can be removed
* `BuiltinType`/`UnionType` have to override `isNullable()` and return `true` when appropriate

With these little changes I believe nullability would be properly covered.

File: src/Symfony/Component/TypeInfo/Type.php (hunk 13 above)
Comment (stof) [reply src/Symfony/Component/TypeInfo/Type.php]: Well, I totally agree that `isA` is weird (maybe it only needs to be renamed). But that's one more reason why removing `is` by suggesting `isA` as a replacement is a very bad idea.

**Comment by stof** [comment]
@skalpa actually, there are more nullable types:

- an intersection type where all its types are nullable
- a union type containing a nullable type (which might not be `null`, as you said yourselves that there are more nullable types, for instance `mixed` or other union types)

**Comment by skalpa** [comment]
> @skalpa actually, there are more nullable types:
> * an intersection type where all its types are nullable
> * a union type containing a nullable type (which might not be `null`, as you said yourselves that there are more nullable types, for instance `mixed` or other union types)

@stof, intersections cannot contain nullable types, only [class/interface names](https://www.php.net/manual/en/language.types.type-system.php), which is logical (nothing could be both an instance of `Stuff` and `null` at the same time). Currently it is possible to create nonsensical intersections that would be invalid in PHP, but I fixed that in #57943.

For unions, yes, I meant "a union that contains a nullable" and not "a union that contains null", but after seeing your comment and checking, I believe it should be forbidden as well. Trying to add `mixed` to a union gives me `PHP Fatal error:  Type mixed can only be used as a standalone type`.

And a union cannot contain another union: this throws an exception (`Type::union()` lets you pass other unions but it merges them to a new one before calling the constructor).

**Comment by stof** [comment]
@skalpa as said above, `mixed` is nullable. And it could be used in an IntersectionType (unless you canonicalize such intersection `mixed&T` to be `T` during instantiation).

And why the PHP syntax enforces types to be written in disjunctive normal form (meaning no union inside intersection), I'm not sure TypeInfo should have this restriction on the created object graph (or it should have proper types enforcing valid usages).
And enforcing canonical forms might make `asNonNullable` more complex because of the fact that turning `mixed` into a non-nullable type turns it into a union.

**Comment by skalpa** [comment]
> @skalpa as said above, `mixed` is nullable. And it could be used in an IntersectionType (unless you canonicalize such intersection `mixed&T` to be `T` during instantiation).
> 

@stof, No, an intersection cannot contain `mixed`. I am sorry to insist, but this is absolutely crucial as it completely changes the semantics of the type if not enforced. In PHP, intersections represent objects. If a value satisfies an intersection, then one must be able to assume that `true === is_object($value)`. And PHP forbids using anything else than class names in intersections.

While the normal form is something else, I really believe we should prevent the creation of nonsensical and/or invalid types that would never be satisfied by any value or would only result in PHP fatal errors if used, as they may probably be indicators of a bug in user code.

If this is not enforced by the component, then as a user I will have to anticipate absolutely impossible cases just because the API allows them and I'd have to protect my code against intersections of scalars, or unions of `mixed` and `never`. ☹️

(even though yes, technically you're right, any intersection is always equivalent to itself & mixed)

> And why the PHP syntax enforces types to be written in disjunctive normal form (meaning no union inside intersection), I'm not sure TypeInfo should have this restriction on the created object graph (or it should have proper types enforcing valid usages). And enforcing canonical forms might make `asNonNullable` more complex because of the fact that turning `mixed` into a non-nullable type turns it into a union.

Some restrictions/enforcements are already there: you are not allowed to add a union to a union, or an intersection to another intersection, which is completely arbitrary if you consider that these could be merged later. But in fact I believe that @mtarld's is the right approach here and am just trying to build on top of it:

IMHO, object graphs should be as canonical as possible. For instance, the fact that unions sort their subtypes is brilliant and without it, to compare `Type::union(Type::int(), Type::string())` to `Type::union(Type::string(), Type::int())` would be near impossible. 

To make this possible requires enforcing the normal form in the `Type` classes afaik. This is actually partly implemented but incomplete, which is what I tried to improve in my own PR.

In return, `TypeFactoryTrait` methods can be used to add some flexibility to types creation and ensure that `Type::union(Type::int(), Type::union(Type::string(), Type::bool()))` is not only allowed but will create the exact same type as `Type::union(Type::bool(), Type::int(), Type::string())`.

If we consider that we should allow any random combination of types, then the current checks should be removed (but I would consider that a step back).

TL;DR (sorry but I really like this component, I talk too much sometimes): I'd like to be able to assume that `echo $type` will always give me a syntactically valid PHP type. We can already use `Type::union()` and `Type::intersection()` to combine composite types in a more flexible way and could add more if necessary.

File: src/Symfony/Bridge/Doctrine/composer.json (hunk 4 above)
Comment (mtarld) [reply src/Symfony/Bridge/Doctrine/composer.json:42]: Indeed, it wasn't, thanks!

**Comment by mtarld** [comment]
Ok, I ended up with the following:
- I renamed `Type::is` to `Type::satisfy` and `Type::isA` to `Type::isIdentifiedBy` so that it is easier to understand
- I added the `CompositeTypeInterface::composedTypesSatisfy` and `WrappingTypeInterface::wrappedTypeSatisfy` methods so that it is easier to define whether we want to check the root, the children or the leaves (you can have a look for an example in `Type::isIdentifiedBy`

I'm still not sure if the `Type::asNonNullable` is needed, as we now have that `NullableType`. I mean, does it make sense to make `mixed` or `null` as non-nullable? To me, the main aim for that method me is to recover `int` in `int|null`, which is covered by `NullableType::getWrappedType`.

Plus, I disallowed the creation of an `UnionType` with `null` as part, and enforced the creation of a `NullableType` instead (see `UnionType::__construct`). But I'm not sure if that's a good idea.

WDYT?

File: src/Symfony/Component/TypeInfo/Type/TemplateType.php (hunk 15 above)
Comment (mtarld) [reply src/Symfony/Component/TypeInfo/Type/TemplateType.php:48]: I didn't know that we had to be that careful with experimental components, I reverted that, thanks

File: src/Symfony/Component/TypeInfo/Type.php (hunk 13 above)
Comment (mtarld) [reply src/Symfony/Component/TypeInfo/Type.php]: You're totally right, that's why I renamed `Type::is` to `Type::satisfy` and `Type::isA` to `Type::isIdentifiedBy` so that it is easier to understand. And I added the `CompositeTypeInterface::composedTypesSatisfy` and `WrappingTypeInterface::wrappedTypeSatisfy` so that it is easier to define whether we want to check the root, the children or the leaves.

File: src/Symfony/Component/TypeInfo/Type.php (hunk 9 above)
Comment (mtarld) [reply src/Symfony/Component/TypeInfo/Type.php]: Great idea yes, I updated the method signature so that it can take a variadic argument

File: src/Symfony/Component/TypeInfo/Type/NullableType.php (hunk 11 above)
Comment (mtarld) [reply src/Symfony/Component/TypeInfo/Type/NullableType.php:42]: As `UnionType` has a variadic argument, yes I think

**Comment by mtarld** [comment]
Moreover, I pulled some ideas from the great work of @skalpa in https://github.com/symfony/symfony/pull/57943 that make types more consistent and make my work on nullable types more easy and relevant.

**Comment by chalasr** [comment]
Rebase needed.
Also we need to move forward on this as the 7.2 feature freeze will come fast making TypeInfo's API stable and there are several open PRs that depend on the API updated here. @stof can you please have another look?

**Comment by xabbuh** [comment]
The failures look related.

**Comment by mtarld** [comment]
These changes are making types generation stricter, and they allowed us to catch a bug: https://github.com/symfony/symfony/pull/58371

File: src/Symfony/Component/TypeInfo/Type.php [hunk 16]
```diff
@@ -25,35 +24,36 @@ abstract class Type implements \Stringable
...
-    abstract public function getBaseType(): BuiltinType|ObjectType;
-
     /**
-     * @param TypeIdentifier|class-string $subject
+     * @param callable(self): bool $specification
      */
-    abstract public function isA(TypeIdentifier|string $subject): bool;
-
-    abstract public function asNonNullable(): self;
+    public function satisfy(callable $specification): bool
```
Comment (chalasr) [inline src/Symfony/Component/TypeInfo/Type.php]: `isSatisfiedBy()` instead?

File: src/Symfony/Component/TypeInfo/Type/NullableType.php (hunk 12 above)
Comment (mtarld) [reply src/Symfony/Component/TypeInfo/Type/NullableType.php:50]: > this type actually composes both this type and the NULL type. It might cause confusion for code not handling properly types implementing both WrappingTypeInterface and IntersectionType.

I'm not sure to understand what you mean here, could you elaborate please?

> And this `getWrappedType` is not even a replacement for `asNonNullable` as nothing prevents the NULL type to be part of the wrapped type.

From now on, the `NullableType` can only be constructed with a non-nullable type, therefore `asNonNullable` is now equal to `getWrappedType`.

File: src/Symfony/Component/TypeInfo/CHANGELOG.md [hunk 17]
```diff
@@ -4,6 +4,13 @@ CHANGELOG
 7.2
 ---
 
+ * Add construction validation for `BackedEnumType`, `CollectionType`, `GenericType`, `IntersectionType`, and `UnionType`
+ * Add `TypeIdentifier::isStandalone`, `TypeIdentifier::isScalar`, and `TypeIdentifier::isBool` methods
+ * Add `WrappingTypeInterface` and `CompositeTypeInterface` type interfaces
+ * Add `NullableType` type class
+ * Rename `Type::isA` to `Type::isIdentifiedBy` and `Type::is` to `Type::satisfy`
```
Comment (chalasr) [inline src/Symfony/Component/TypeInfo/CHANGELOG.md]:
```suggestion
 * Rename `Type::isA()` to `Type::isIdentifiedBy()` and `Type::is()` to `Type::isSatisfiedBy()`
```
same for other referenced methods about the `()` suffix

**Review by chalasr — CHANGES_REQUESTED** [review]
Please add some phpdoc to `isidentifiedBy()` and `isSatisfiedBy()` explaining what they are supposed to do and how they differ, so that one doesn't have to read the implementation for that

File: src/Symfony/Component/TypeInfo/CHANGELOG.md [hunk 18]
```diff
@@ -4,6 +4,13 @@ CHANGELOG
 7.2
 ---
 
+ * Add construction validation for `BackedEnumType`, `CollectionType`, `GenericType`, `IntersectionType`, and `UnionType`
+ * Add `TypeIdentifier::isStandalone`, `TypeIdentifier::isScalar`, and `TypeIdentifier::isBool` methods
+ * Add `WrappingTypeInterface` and `CompositeTypeInterface` type interfaces
+ * Add `NullableType` type class
+ * Rename `Type::isA` to `Type::isIdentifiedBy` and `Type::is` to `Type::isSatisfiedBy`
+ * Remove `Type::getBaseType`, `Type::asNonNullable` and `Type::__call` methods
```
Comment (chalasr) [inline src/Symfony/Component/TypeInfo/CHANGELOG.md]:
```suggestion
 * Rename `Type::isA()` to `Type::isIdentifiedBy()` and `Type::is()` to `Type::isSatisfiedBy()`
 * Remove `Type::getBaseType()`, `Type::asNonNullable()` and `Type::__call()` methods
```

**Review by chalasr — APPROVED** [review]
Better API

**Comment by fabpot** [comment]
It looks like the tests are still failing.

**Comment by mtarld** [comment]
The high-deps build is failing because this PR purposely contains BC breaks as the component is experimental in 7.2. Failures will go away once this is merged.

**Comment by chalasr** [comment]
Indeed, tests will pass once merged. Then dependant components' tests will start failing on 7.1 as they will install type-info 7.2 so we will have to forbid it by adding conflict rules.

**Comment by xabbuh** [comment]
The issue is that the conflict will lead to the fact that Composer may not install the latest 7.1 version of said component because of the conflict when 7.2 of the TypeInfo component is installed as we already have releases claiming to be compatible with TypeInfo 7.2. So the damage is already done. The only solution I see is to make the PropertyInfo component compatible with TypeInfo 7.2 as a bugfix in 7.1.

In the future we need to be careful when adding requirements for experimental components and do not allow to install new minor versions of the from the beginning.

**Comment by stof** [comment]
@xabbuh a way to avoid that breakage is to make type-info conflict with older versions of those components, to workaround the fact that we used semver constraints for an experimental package not guaranteeing semver yet.

**Comment by chalasr** [comment]
Sounds good

**Comment by xabbuh** [comment]
Indeed, @stof's suggestion is a good idea that we can use in this PR.

**Comment by mtarld** [comment]
Yes, that a great idea, I just pushed that fix.

**Comment by chalasr** [comment]
@symfony/mergers Additional review needed to move forward before the component is considered stable.

**Review by nicolas-grekas — APPROVED** [review]
Let's give this a try :)

**Comment by nicolas-grekas** [comment]
Thank you @mtarld.
